### PR TITLE
Handling exceptions for sync operations (fixes #59).

### LINKED ISSUE #59: Exceptions for sync operations.
Many of the sync operations are documented as returning one of the following exceptions:

- `ResourceLockedException`
- `ResourceNotFoundException`
- `DeclareException`

We can detect these conditions using the AMQP replyCode in channel/connection close frames, but the only way to know for sure that they respond to a given request is when using a managed channel (because nothing else can be happening on the channel).

We need to actually implement the code that catches the `AmqpException` (see #15) and wraps it in one of the three listed above, but this should only be done for auto-managed channels, otherwise we should let the normal connection/channel exception bubble.

### PR DESCRIPTION
**Pull request by koden-km**
Not all tests for Amqp091Consumer.php have been implemented.
There was no test file for that class yet and it seems to have a different code style than the Exchange and Queue classes.
Is it an older implementation that will be refactored?
Should i go ahead and try write the tests for it now anyway?

### CODE REVIEW

File: test/suite/v091/Amqp091ConsumerTest.php [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        $this->subject = new Amqp091Consumer(
+            $this->broker->mock(),
+            $this->queue->mock(),
+            $this->options,
+            $this->features,
+            '',
+            $this->channel->mock()
+        );
+
+        $exception = new LogicException(
```
Comment (jmalloc) [inline test/suite/v091/Amqp091ConsumerTest.php]: Unused `$exception` variable.

**Comment by jmalloc** [comment]
You don't need to write tests for all of `Amqp091Consumer`, it *is* quite different to exchange and queue, but it just... is.. different :P  It's not old code, I just haven't written tests for it yet, but it should be mostly stable.  I'm working on it's counterpart, the publishers, in the `feature/publish` branch atm.

**Comment by koden-km** [comment]
I wasn't sure if i covered the right exceptions in `Amqp091Exchange::create()`.
It lists two types of `DeclareException`'s:
```
     * @throws DeclareException    [via promise] The exchange already exists with different a type or options.
     * @throws DeclareException    [via promise] A reserved exchange name was used.
```
But only one seemed to match the current `DeclareException` exchange create function: `DeclareException::exchangeTypeOrOptionMismatch()`.
Should a new exchange create function be made for the "A reserved exchange name was used." error?

**Comment by jmalloc** [comment]
Ah yeah, that one needs a new static method, and should be thrown if the exchange name starts with `amq.`

**Comment by jmalloc** [comment]
By the way, you might want to [check this out](http://opensoul.org/2012/11/09/convert-a-github-issue-into-a-pull-request/) for these PRs that are for a single existing GH issue.

**Comment by koden-km** [comment]
This should be ready to merge now if you are happy with the name of that new DeclareException method.
The consumer doesn't have all tests yet, but as you said that should be done in a separate task.

File: src/Exception/DeclareException.php [hunk 2]
```diff
@@ -43,6 +43,29 @@ public static function exchangeTypeOrOptionMismatch(
...
+     *
+     * @return DeclareException
+     */
+    public static function exchangeNameIsReserved(
+        $name,
+        Exception $previous = null
+    ) {
+        return new self(
+            sprintf(
+                'Failed to declare exchange "%s" name is reserved.',
```
Comment (jmalloc) [inline src/Exception/DeclareException.php]: `'Failed to declare exchange "%s", names beginning with "amp." are reserved.'`

File: src-internal/v091/Amqp091Exchange.php [hunk 3]
```diff
@@ -363,10 +366,25 @@ function ($exception, $brokerChannel) use ($channel, $passive, $deferred) {
                         function ($exception, $frame) use ($brokerChannel, $channel, $deferred) {
                             if (!$channel) {
                                 $this->broker->releaseChannel($brokerChannel);
+
+                                if ($exception instanceof AmqpException) {
+                                    if (substr($this->name, 0, strlen('amqp.')) === 'amqp.') {
```
Comment (jmalloc) [inline src-internal/v091/Amqp091Exchange.php]: This needs to be `amq.`, there is no `p` (https://www.rabbitmq.com/tutorials/amqp-concepts.html#exchanges)

**Comment by koden-km** [comment]
Fixed as suggested.

**Pull request merged**